### ISSUE TITLE
Support Starting AWS Database Migration Service Replication Task

### ISSUE THREAD

**Issue by shelley-jean-baker** [terraform-providers/terraform-provider-aws#2083]
Add support for starting AWS Database Migration Service `full-load-and-cdc` replication task after successful creation of the [`aws_dms_replication_task`](https://www.terraform.io/docs/providers/aws/r/dms_replication_task.html) resource.

The AWS DMS console provides an option to "[Start task on create](http://docs.aws.amazon.com/dms/latest/sbs/CHAP_On-PremOracle2Aurora.Steps.CreateTask.html)", and the AWS CLI provides a [`start-replication-task`](http://docs.aws.amazon.com/cli/latest/reference/dms/start-replication-task.html) command, but there doesn't appear to be any similar option in the [Terraform resources](https://www.terraform.io/docs/providers/aws/). The [`aws_dms_replication_task`](https://www.terraform.io/docs/providers/aws/r/dms_replication_task.html) provides the [`cdc_start_time`](https://www.terraform.io/docs/providers/aws/r/dms_replication_task.html#cdc_start_time) argument, but this seems to apply only to `cdc` tasks. Optional support to automatically start the task after successful creation would be beneficial.

Related questions on [StackOverflow](https://stackoverflow.com/questions/46920760/starting-aws-dms-replication-task-in-terraform) and the [Terraform Google Group](https://groups.google.com/forum/#!topic/terraform-tool/Cavx7e8PUJI).

**Comment by guidorugo** [comment]
Any news about this ? This is really helpful

**Comment by leeuw471** [comment]
> Any news about this ? This is really helpful

Indeed, would really love this option

**Comment by tonycodurance** [comment]
> Any news about this ? This is really helpful

Sounds like a great idea!

**Comment by bm1549** [comment]
Looking at the [AWS API Docs](https://docs.aws.amazon.com/dms/latest/APIReference/API_CreateReplicationTask.html) and [AWS Cloudformation Docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-dms-replicationtask.html) it doesn't seem like something that AWS supports 😞

**Comment by leeuw471** [comment]
Wouldn't that be https://docs.aws.amazon.com/dms/latest/APIReference/API_StartReplicationTask.html though?

**Comment by bm1549** [comment]
Yes, but the Terraform aws provider code I looked at was a single api call per resource crud operation, so I'm not sure they'd implement this

**Comment by raymondwclark** [comment]
Any updates on this?  Would be a great thing to have.

**Comment by dennisotugo** [comment]
Bump

**Comment by rafaelrsantosti** [comment]
Any update ?

**Comment by luizm** [comment]
A workaround I've been used:

```sh
resource "null_resource" "example" {
  provisioner "local-exec" {
    command = "aws dms start-replication-task --start-replication-task-type start-replication --replication-task-arn ${aws_dms_replication_task.example.replication_task_arn}"
  }
```

**Comment by ceckbold** [comment]
It does seem odd to have the ability to provision the DMS resources but no way to start the actual migration process. 

+1 luizm for the workaround.

**Comment by awahnteh** [comment]
Very odd indeed

**Comment by peteches** [comment]
Small improvement on that workaround @luizm posted:

`resource "null_resource" "start_replicating" {
  triggers = {
    dms_task_arn = aws_dms_replication_task.example.replication_task_arn
  }
  provisioner "local-exec" {
    when    = create
    command = "aws dms start-replication-task --start-replication-task-type start-replication --replication-task-arn ${self.triggers["dms_task_arn"]}"
  }

}
`

I get deprecation warnings when interpolating resources and variables inside the provisioner command.  Which ties in nicely with another issue, that if the DMS taks it tries to start is stopped, or deleted the null_provisioner won't restart it.

By triggering on the dms task's arn we can solve both problems at once. the trigger takes care of making suire the null_resource is rebuilt when needed but also makes the value available in the self object for the command to use without having to worry about deprecation warnings.

**Comment by zetas** [comment]
We make heavy use of DMS is there any progress on this working? Using the local-exec workaround is adding quite a bit of flakiness to our terraform builds and destroys, a proper solution would be nice.

**Comment by felipeanchieta** [comment]
Any news?

**Comment by sandeepbabylon** [comment]
Any news on this feature ?

**Comment by ddvdozuki** [comment]
Any updates on this?

**Comment by darkplayground** [comment]
Any update to this? Still having tons of issues starting and stopping DMS tasks during create and destroy. Null resource provisioners are a stop-gap at best, they introduce all kinds of race conditions. If the API is a little slow your delete fails because the task hasn't fully stopped.

**Comment by ddvdozuki** [comment]
I'm reticent to post this as It might discourage the maintainers from actually adding support but seeing as how this post is over 4 years old I might as well put our solution here. We've been running this in production deploys for our customers for months now and It's never had an issue yet. A combination of a Kubernetes job to start the task after the app has finished provisioning and a null_resource delete-time provisioner with a shell script to stop the task.

```hcl
resource "kubernetes_job" "dms_start" {
  count = var.enable_bi ? 1 : 0

  depends_on = [helm_release.replicated]

  metadata {
    name = "dms-start"
  }
  spec {
    template {
      metadata {}
      spec {
        container {
          name  = "dms-start"
          image = "bearengineer/awscli-kubectl:latest"
          command = [
            "/bin/sh",
            "-c",
            "kubectl -n ${coalesce([for i, v in data.kubernetes_all_namespaces.allns.namespaces : try(regexall("replicated\\-.*", v)[0], "")]...)} wait deploy/app-deployment --for condition=available --timeout=1200s && aws dms start-replication-task --start-replication-task-type start-replication --replication-task-arn ${var.dms_task_arn} --region ${data.aws_region.current.name}"
          ]
        }
        restart_policy = "Never"
      }
    }
    backoff_limit = 1
    completions   = 1
  }
  wait_for_completion = true

  timeouts {
    create = "20m"
  }
}
```
We use coalesce to pull the randomly-generated namespace by replicated (the deploy tool our app uses) and wait for the app deployment to become ready before starting the DMS task. This way when DMS starts it pulls the entire database and not just whatever is there on new RDS creation. We've had issues with DMS picking up new table creations in MySQL to MySQL replication tasks.

Then we have our null_resource delete provisioner to handle the opposite direction:
```hcl
resource "null_resource" "replication_control" {
  count = var.enable_bi ? 1 : 0

  triggers = {
    dms_task_arn        = aws_dms_replication_task.this[0].replication_task_arn,
    source_endpoint_arn = aws_dms_endpoint.source[0].endpoint_arn,
    target_endpoint_arn = aws_dms_endpoint.target[0].endpoint_arn,
    aws_region          = data.aws_region.current.name,
    aws_profile         = var.aws_profile
  }

  provisioner "local-exec" {
    when = destroy
    # The sleep fixes a race condition where the dms task isn't fully stopped in time.
    command = "/usr/bin/env bash ./util/dms-stop.sh ${self.triggers["dms_task_arn"]} ${self.triggers["aws_region"]} ${self.triggers["aws_profile"]}"
  }
}
```

And the `dms-stop.sh` script:
```bash
#!/usr/bin/env bash

TRIGGER="$1"
AWS_REGION="$2"
AWS_PROFILE="$3"

function getDMSStatus() {
  aws dms describe-replication-tasks --filter Name=replication-task-arn,Values="$TRIGGER" --without-settings --region "$AWS_REGION" --profile "$AWS_PROFILE" |jq --raw-output '.[][0]["Status"]'
}

function stopDMS() {
  local STATUS

  STATUS=$(getDMSStatus)
   if [[ "$STATUS" == "running" ]]; then
     echo -e "Stopping DMS Task..."
     aws dms stop-replication-task --replication-task-arn "$TRIGGER" --region "$AWS_REGION" --profile "$AWS_PROFILE" > /dev/null
     waitforStoppedDMS
   else
     echo -e "DMS Already Stopped."
   fi
}

function waitforStoppedDMS() {
  echo "Waiting for the DMS task to stop..."

  local STATUS

  STATUS=$(getDMSStatus)
  while [[ "$STATUS" == "stopping" ]]; do
    sleep 10
    STATUS=$(getDMSStatus)
  done

  echo -e "DMS Task Stopped Successfully."
}

stopDMS

```

Hopefully this helps anyone else that suffers with supporting DMS in terraform. Also feel free to let me know if you see any improvements to this system. I'm always looking to better myself 👍
